File: lib/src/features/pitches/domain/entities/pitch.dart
import 'package:equatable/equatable.dart';

class Pitch extends Equatable {
  final String id;
  final String name;
  final String description;
  final String ownerId;
  final String sportType;
  final PitchLocation location;
  final double pricePerHour;
  final List<String> amenities;
  final List<String> images;
  final double rating;
  final int reviewsCount;
  final bool isAvailable;
  final String openingTime;
  final String closingTime;
  final int slotDurationMinutes;
  final double morningPrice;
  final double nightPrice;

  const Pitch({
    required this.id,
    required this.name,
    required this.description,
    required this.ownerId,
    required this.sportType,
    required this.location,
    required this.pricePerHour,
    required this.amenities,
    required this.images,
    required this.rating,
    required this.reviewsCount,
    required this.isAvailable,
    required this.openingTime,
    required this.closingTime,
    required this.slotDurationMinutes,
    required this.morningPrice,
    required this.nightPrice,
  });

  factory Pitch.empty() => const Pitch(
    id: '',
    name: 'Pitch Name Placeholder',
    description: '',
    ownerId: '',
    sportType: '',
    location: PitchLocation(
      address: 'Address',
      city: 'City',
      country: 'Country',
      latitude: 0,
      longitude: 0,
    ),
    pricePerHour: 0,
    amenities: [],
    images: [],
    rating: 0,
    reviewsCount: 0,
    isAvailable: true,
    openingTime: '',
    closingTime: '',
    slotDurationMinutes: 60,
    morningPrice: 0,
    nightPrice: 0,
  );

  String get imageUrl => images.isNotEmpty ? images.first : '';

  @override
  List<Object?> get props => [
    id,
    name,
    description,
    ownerId,
    sportType,
    location,
    pricePerHour,
    amenities,
    images,
    rating,
    reviewsCount,
    isAvailable,
    openingTime,
    closingTime,
    slotDurationMinutes,
    morningPrice,
    nightPrice,
  ];
}

class PitchLocation extends Equatable {
  final String address;
  final String city;
  final String country;
  final double latitude;
  final double longitude;

  const PitchLocation({
    required this.address,
    required this.city,
    required this.country,
    required this.latitude,
    required this.longitude,
  });

  String get fullAddress => '$address, $city, $country';

  @override
  List<Object?> get props => [address, city, country, latitude, longitude];
}

File: e7gz_backend/src/models/Pitch.js
import mongoose from 'mongoose';

const pitchSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: [true, 'Pitch name is required'],
      trim: true,
    },
    description: {
      type: String,
      trim: true,
    },
    ownerId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    sportType: {
      type: String,
      enum: ['football', 'basketball', 'tennis', 'padel', 'volleyball', 'multi'],
      default: 'football',
    },
    location: {
      address: { type: String, required: true },
      city: { type: String, required: true },
      country: { type: String, default: 'Egypt' },
      // GeoJSON for geospatial queries
      coordinates: {
        type: {
          type: String,
          enum: ['Point'],
          default: 'Point',
        },
        coordinates: {
          type: [Number], // [longitude, latitude]
          required: true,
        },
      },
    },
    pricePerHour: {
      type: Number,
      required: [true, 'Price per hour is required'],
      min: 0,
    },
    amenities: {
      type: [String],
      default: [],
    },
    images: {
      type: [String],
      default: [],
    },
    rating: {
      type: Number,
      default: 0,
      min: 0,
      max: 5,
    },
    reviewsCount: {
      type: Number,
      default: 0,
    },
    isAvailable: {
      type: Boolean,
      default: true,
    },
    openingTime: {
      type: String,
      default: '06:00', // HH:mm
    },
    closingTime: {
      type: String,
      default: '24:00',
    },
    slotDurationMinutes: {
      type: Number,
      default: 60,
    },
  },
  { timestamps: true }
);

// Create 2dsphere index for geo queries
pitchSchema.index({ 'location.coordinates': '2dsphere' });

export default mongoose.model('Pitch', pitchSchema);
